src/nbl/video/CVulkanLogicalDevice.cpp

```diff
@@ -0,0 +1,100 @@
+#include "CVulkanLogicalDevice.h"
+
+#include "nbl/video/CVulkanPhysicalDevice.h"
+
+namespace nbl::video
+{
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateDeviceLocalMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (getDeviceLocalGPUMemoryReqs().memoryHeapLocation != additionalReqs.memoryHeapLocation)
+        return nullptr;
+
+    return allocateGPUMemory(additionalReqs, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+}
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateSpilloverMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (getSpilloverGPUMemoryReqs().memoryHeapLocation != additionalReqs.memoryHeapLocation)
+        return nullptr;
+
+    return allocateGPUMemory(additionalReqs, static_cast<VkMemoryPropertyFlags>(0u));
+}
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateUpStreamingMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (getUpStreamingMemoryReqs().memoryHeapLocation != additionalReqs.memoryHeapLocation)
+        return nullptr;
+
+    return allocateGPUMemory(additionalReqs, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+}
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateDownStreamingMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (getDownStreamingMemoryReqs().memoryHeapLocation != additionalReqs.memoryHeapLocation)
+        return nullptr;
+
+    return allocateGPUMemory(additionalReqs, static_cast<VkMemoryPropertyFlags>(0u));
+}
+
+core::smart_refctd_ptr<IDriverMemoryAllocation> CVulkanLogicalDevice::allocateCPUSideGPUVisibleMemory(
+    const IDriverMemoryBacked::SDriverMemoryRequirements& additionalReqs)
+{
+    if (getCPUSideGPUVisibleGPUMemoryReqs().memoryHeapLocation != additionalReqs.memoryHeapLocation)
+        return nullptr;
+
+    return allocateGPUMemory(additionalReqs, static_cast<VkMemoryPropertyFlags>(0u));
```

Comment:
ok so the correct way to handle these is to upgrade the `additionalReqs.memoryHeapLocation` examples:
- Device Local = always upgrade to Device Local
- Spillover,Streaming = max(getSpilloverMemoryReqs().memoryHeapLocation,additionalReqs.memoryHeapLocation), but fail if the spillover reqs say not-device-local but additional reqs require device local
- CPUSideGPUVisible = always non-device local, fail if additionalReqs require device local

then you can take the maximum for alignment and size

bitwise OR the mapping capabilities

bitwise AND the `memoryTypeBits`

Reply:
1. > then you can take the maximum for alignment and size
    > 
    > bitwise AND the memoryTypeBits

Do you still want these even if `get**GPUMemoryReqs` always return:
`size == 0`
`alignment == 0`
`memoryTypeBits == 0xffffffff`
?

2. > bitwise OR the mapping capabilities

Doing this might result in conflicting bits being set (`getDeviceLocalGPUMemoryReqs` might set cannot map, user might set can map, for example). In that case, I would have to detect when a conflict occurs and clear the set bits. It might be much more effort (and bug-prone) than simply overwriting with `additionalReqs.mappingCapability`. Worth it?

Reply:
> Do you still want these even if get**GPUMemoryReqs always return:

it will clearly show what's happening.

> Doing this might result in conflicting bits being set (getDeviceLocalGPUMemoryReqs might set cannot map, user might set can map, for example). In that case, I would have to detect when a conflict occurs and clear the set bits. It might be much more effort (and bug-prone) than simply overwriting with additionalReqs.mappingCapability. Worth it?

the mapping Capability usually wouldn't be set in `additionalReqs` as you've shown in 
https://github.com/Devsh-Graphics-Programming/Nabla/pull/160#discussion_r700693356

so if anything it would make more sense to not take `additionalReqs.mapppingCapability` into account.

Reply:
> the mapping Capability usually wouldn't be set in additionalReqs as you've shown in
#160 (comment)
> 
> so if anything it would make more sense to not take additionalReqs.mapppingCapability into account.

The opposite is actually true. In the (now defunct) code referenced in here https://github.com/Devsh-Graphics-Programming/Nabla/pull/160#discussion_r700693356, the preferred value of `mappingCapability` comes in through the `additionalReqs` param not by the result of `static inline` functions.

Reply:
obsolete, will get cleaned up into something more sane.